streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java

```diff
@@ -34,13 +34,30 @@
      * @param streamName The given name will be printed.
      */
     public PrintForeachAction(final PrintWriter printWriter, final String streamName) {
+        this(printWriter, null, streamName);
+    }
+
+    /**
+     * Print customized output with given writer. The PrintWriter can be null in order to
+     * distinguish between {@code System.out} and the others. If the PrintWriter is {@code PrintWriter(System.out)},
+     * then it would close {@code System.out} output stream.
+     * <p>
+     * Afterall, not to pass in {@code PrintWriter(System.out)} but {@code null} instead.
+     *
+     * @param printWriter Use {@code System.out.println} if {@code null}.
+     * @param mapper The mapper which can allow user to customize output will be printed.
+     * @param streamName The given name will be printed.
+     */
+    public PrintForeachAction(final PrintWriter printWriter, KeyValueMapper<? super K, ? super V, String> mapper, final String streamName) {
         this.printWriter = printWriter;
```

Comment:
I think @dguy 's point is that, it is better to keep the `null` checks only in one function, and that function would better be as higher-level in the call trace (e.g. on the callers rather than callee). For this specific class, we can consolidate the `null` checks only in the `KStreamImpl`, and pass the default `mapper` / `label` / `writer` etc when the passed values are null, then we do not need to remember doing such further checkings in the lower level classes like `PrintForeachAction`.

So I'd say we can at least try enforcing such a pattern in this PR and gradually improve other functions other than `print` in future PRs.